From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Adallei <Adallei@outlook.com>
Date: Sun, 16 Aug 2026 09:51:47 +0800
Subject: [PATCH] Backport config buffer joins to world

Backport from PaperMC

This patch buffers the number of logins which will attempt to join
the world per tick, this attempts to reduce the impact that join floods
has on the server

diff --git a/src/main/java/com/adallei/alr/AlrConfig.java b/src/main/java/com/adallei/alr/AlrConfig.java
index 779895279170d7c63702ae3c4e3bf609c837c0aa..8a24ed07d182c0109e888b280019370c31996028 100644
--- a/src/main/java/com/adallei/alr/AlrConfig.java
+++ b/src/main/java/com/adallei/alr/AlrConfig.java
@@ -250,4 +250,9 @@ public class AlrConfig {
             DROP;
         }
     }
+
+        public static int maxJoinsPerTick;
+        private static void maxJoinsPerTick() {
+            maxJoinsPerTick = getInt("settings.max-joins-per-tick", 3);
+            }
 }
diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
index 668ab0d3bbc85d961a4748095455a2026a869107..01fa40071a47ef12967b58645041e1b13a3ea814 100644
--- a/src/main/java/net/minecraft/server/LoginListener.java
+++ b/src/main/java/net/minecraft/server/LoginListener.java
@@ -39,7 +39,7 @@ public class LoginListener implements PacketLoginInListener, ITickable {
     private final byte[] e = new byte[4];
     private final MinecraftServer server;
     public final NetworkManager networkManager;
-    private LoginListener.EnumProtocolState g;
+    private LoginListener.EnumProtocolState g; public final LoginListener.EnumProtocolState getLoginState() { return this.g; }; // Paper - OBFHELPER
     private int h;
     private GameProfile i; private void setGameProfile(GameProfile profile) { i = profile; } private GameProfile getGameProfile() { return i; } // Paper - OBFHELPER
     private final String j;
diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index 65ad94acb6624383808df592725e78ccbeaebe16..3766739293d8c7749b2fd36987cdab233edd40da 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -473,10 +473,27 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
     }
     // Paper end
 
+    private static final int MAX_PER_TICK = com.adallei.alr.AlrConfig.maxJoinsPerTick; // Paper
+    private static int joinAttemptsThisTick; // Paper
+    private static int currTick; // Paper
     public void a() {
         this.m();
-        if (this.m instanceof ITickable) {
-            ((ITickable) this.m).e();
+        // Paper start
+        if (currTick != MinecraftServer.currentTick) {
+            currTick = MinecraftServer.currentTick;
+            joinAttemptsThisTick = 0;
+        }
+        // Paper end
+        if (this.m instanceof LoginListener) {
+            LoginListener loginListener = (LoginListener) this.m;
+            if (loginListener.getLoginState() != LoginListener.EnumProtocolState.READY_TO_ACCEPT  // Paper
+                    || (joinAttemptsThisTick++ < MAX_PER_TICK)) { // Paper - limit the number of joins which can be processed each tick
+                loginListener.e();
+            }
+        }
+
+        if (this.m instanceof PlayerConnection) {
+            ((PlayerConnection)this.m).e();
         }
 
         if (this.channel != null) {
